### PR TITLE
Refactor InternalEngine's index/delete flow for better clarity

### PR DESCRIPTION
**Pull request by bleskes**
The InternalEngine Index/Delete methods (plus satellites like version loading from Lucene) have accumulated some cruft over the years making it hard to clearly the code flows for various use cases (primary indexing/recovery/replicas etc). This PR refactors those methods for better readability. As a follow up we intend to take certain parts of these method and extract them to another help methods to improve things even more. This will be done as a follow up.

To support the refactoring I have considerably beefed up  the versioning tests.

This PR is a spin-off from #23543 , which made it clear this is needed.

### CODE REVIEW

**Comment by bleskes** [comment]
@jasontedor This is the same code you already reviewed but without the seq no logic. Since you already LGTMed so I didn't ask for your review. Of course, feel free to review anyway if you want.

File: core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java [hunk 1]
```diff
@@ -51,49 +51,52 @@
     private final TermsEnum termsEnum;
     /** _version data */
     private final NumericDocValues versions;
+
     /** Reused for iteration (when the term exists) */
     private PostingsEnum docsEnum;
 
+    private final Object readerKey;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java]: this reader key is only used for asserts. can we make sure it's null if asserts are not enabled?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -402,56 +403,37 @@ public GetResult get(Get get, Function<String, Searcher> searcherFactory) throws
...
-            } else {
-                /* Version conflicts during recovery and on replicas are normal due to asynchronous execution; as such, we should return a
-                 * successful result.*/
-                return true;
+    /** resolves the current version of the document, returning null if not found */
+    private VersionValue resolveDocVersion(final Operation op) throws IOException {
+        assert incrementVersionLookup();
+        VersionValue versionValue = versionMap.getUnderLock(op.uid());
+        if (versionValue == null) {
+            assert incrementIndexVersionLookup();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: message?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -402,56 +403,37 @@ public GetResult get(Get get, Function<String, Searcher> searcherFactory) throws
...
-
-        if (op.versionType().isVersionConflictForWrites(currentVersion, expectedVersion, deleted)) {
-            if (op.origin() == Operation.Origin.PRIMARY) {
-                // fatal version conflict
-                throw new VersionConflictEngineException(
-                        shardId,
-                        op.type(),
-                        op.id(),
-                        op.versionType().explainConflictForWrites(currentVersion, expectedVersion, deleted));
+    enum LuceneOpStatus {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: the enum values make no sense in the context of the name. I wonder if we should call it `OperationAge` or something like this.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -402,56 +403,37 @@ public GetResult get(Get get, Function<String, Searcher> searcherFactory) throws
...
         }
+        return versionValue;
     }
 
-    private long checkDeletedAndGCed(VersionValue versionValue) {
-        long currentVersion;
-        if (engineConfig.isEnableGcDeletes() && versionValue.delete() && (engineConfig.getThreadPool().relativeTimeInMillis() - versionValue.time()) > getGcDeletesInMillis()) {
-            currentVersion = Versions.NOT_FOUND; // deleted, and GC
+    private LuceneOpStatus checkLuceneOpStatusBasedOnVersions(final Operation op) throws IOException {
+        assert op.version() >= 0;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can we get the actual version in the message?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 5]
```diff
@@ -543,62 +534,95 @@ public IndexResult index(Index index) throws IOException {
...
-                }
-                final long expectedVersion = index.version();
-                Optional<IndexResult> resultOnVersionConflict;
-                try {
-                    final boolean isVersionConflict = checkVersionConflict(index, currentVersion, expectedVersion, deleted);
-                    resultOnVersionConflict = isVersionConflict ? Optional.of(new IndexResult(currentVersion, index.seqNo(), false))
-                        : Optional.empty();
-                } catch (IllegalArgumentException | VersionConflictEngineException ex) {
-                    resultOnVersionConflict = Optional.of(new IndexResult(ex, currentVersion, index.seqNo()));
+                } else if (canOptimizeAddDocument && mayHaveBeenIndexedBefore(index) == false) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can we fold this into an `else {` and start again with `if` in there with a comment that we are now on a replica? it would read cleaner

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 6]
```diff
@@ -543,62 +534,95 @@ public IndexResult index(Index index) throws IOException {
...
+                    assert index.version() == 1L : "can optimize on replicas but incoming version is [" + index.version() + "]";
+                    seqNoForIndexing = index.seqNo();
+                    versionForIndexing = index.version();
+                    performOperation = true;
+                } else {
+                    // drop out of order operations
+                    assert index.versionType().versionTypeForReplicationAndRecovery() == index.versionType() :
+                        "resolving out of order delivery based on versioning but version type isn't fit for it. got ["
+                            + index.versionType() + "]";
+                    final LuceneOpStatus luceneOpStatus = checkLuceneOpStatusBasedOnVersions(index);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: the semantics of this method are weird. I pass in an operation and it returns OLDER if the given operation has a higher version? I would have expected the opposite. This semantics also make the rest of this method hard to read since it has to negate the return values in 2/3 of the cases. I think you should flip it.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 7]
```diff
@@ -610,24 +634,27 @@ public IndexResult index(Index index) throws IOException {
         }
     }
 
-    private IndexResult indexIntoLucene(Index index, long seqNo, long currentVersion, boolean deleted, boolean forceUpdateDocument, boolean canOptimizeAddDocument, long expectedVersion) throws IOException {
+    private IndexResult indexIntoLucene(Index index, long seqNo, long newVersion, boolean markDocAsCreated,
+                                        boolean useLuceneUpdateDocument)
+        throws IOException {
+        assertSequenceNumberBeforeIndexing(index.origin(), seqNo);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: only execute this if assertions are enabled?!

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 8]
```diff
@@ -640,18 +667,19 @@ private IndexResult indexIntoLucene(Index index, long seqNo, long currentVersion
                  * `indexWriter.getTragicException()` will be null otherwise we have to rethrow and treat it as fatal or rather
                  * non-document failure
                  */
-                return new IndexResult(ex, currentVersion, index.seqNo());
+                return new IndexResult(ex, Versions.MATCH_ANY, index.seqNo());
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: oh why did you change this to MATCH_ANY?... worth a comment?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 9]
```diff
@@ -640,18 +667,19 @@ private IndexResult indexIntoLucene(Index index, long seqNo, long currentVersion
...
             } else {
                 throw ex;
             }
         }
     }
 
-    private boolean isForceUpdateDocument(Index index) {
-        boolean forceUpdateDocument;
+    private boolean mayHaveBeenIndexedBefore(Index index) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: maybe a javadoc for this method?

**Review by s1monw — CHANGES_REQUESTED** [review]
I left some comments. I have to admit I am not sure it improves the readability of the engine. It rather feels like it make it more complicated with more methods without clear naming.

**Comment by jasontedor** [comment]
> This is the same code you already reviewed but without the seq no logic. Since you already LGTMed so I didn't ask for your review.

I didn't review it, I saw it while it was a work in progress. I have not LGTMed the previous PR, I would like to review this PR.

**Comment by bleskes** [comment]
> I didn't review it, I saw it while it was a work in progress. I have not LGTMed the previous PR, I would like to review this PR.

@jasontedor I'm very sorry and have no idea what made me think you did. As said before, your review is more than welcome (and it seems needed as it didn't do it before).

File: core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java [hunk 10]
```diff
@@ -51,49 +51,52 @@
...
-                termsEnum = terms.iterator();
-                assert termsEnum != null;
-                versions = reader.getNumericDocValues(VersionFieldMapper.NAME);
-                assert versions != null;
-            }
-        }
-
-        this.versions = versions;
-        this.termsEnum = termsEnum;
+        Terms terms = fields.terms(UidFieldMapper.NAME);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java]: That's a good question. The class was introduced that way with no explanation. I thought that it had to do with BWC where we made transitions into how we store UIDs. I wonder when we can end up with an empty reader. I'll discuss this with @jpountz

File: core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java]: good one.

File: core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java [hunk 11]
```diff
@@ -51,49 +51,52 @@
...
+        assert termsEnum != null;
+        versions = reader.getNumericDocValues(VersionFieldMapper.NAME);
+        assert versions != null;
+        readerKey = reader.getCoreCacheKey();
     }
 
     /** Return null if id is not found. */
-    public DocIdAndVersion lookup(BytesRef id, Bits liveDocs, LeafReaderContext context) throws IOException {
+    public DocIdAndVersion lookupVersion(BytesRef id, Bits liveDocs, LeafReaderContext context) throws IOException {
+        assert context.reader().getCoreCacheKey().equals(readerKey);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java]: added.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: The name comes from the following line of thought - maybe it helps explain it (I'm fine with any other names) - operations on replicas (where this is relevant) are always stored in the translog. They are added to lucene only if the document version *in lucene* is older than the incoming one. The enum is supposed to reflect the status of the document in lucene. Does this help? any suggestion as to how to name it to reflect it only relates the lucene index?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: this code is supposed to run only under assertions. I'll add comments.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: yep

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 5 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: sure.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: How about renaming the method and enum to lucene doc status? I'll change that and see if it makes more sense for you.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 7 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: oops. thx.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 9 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: added

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 8 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I don't have that information anymore here, and in the case of failure, we don't use it anyway. I'll comment.

File: core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java (hunk 10 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java]: I double checked with Adrien and this is OK (and is something he wanted to do for a long time).

**Comment by bleskes** [comment]
@s1monw I pushed a commit that address most of your feedback.

re 
>  I have to admit I am not sure it improves the readability of the engine. It rather feels like it make it more complicated with more methods without clear naming.

Fair enough. This is subjective. I think things will be clearer with the follow up change we intended to make (move some the long code into methods that return a struct). Maybe I should do this now within this change so we can see how it looks? @jasontedor indicated he would also prefer to see the end result rather than review this intermediate step (the diff is too big anyway).

**Comment by s1monw** [comment]
> @jasontedor indicated he would also prefer to see the end result rather than review this intermediate step (the diff is too big anyway).

++

**Comment by bleskes** [comment]
@jasontedor @s1monw I pushed ahead and added helper methods. I think it looks much better but this is subjective. I will probably do another run and polish things more but I think it's ready for you. LMKWYT

File: core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java [hunk 12]
```diff
@@ -51,49 +51,71 @@
...
+        if (termsEnum == null) {
+            throw new IllegalArgumentException("reader misses the [" + UidFieldMapper.NAME +
+                "] field");
+        }
+        versions = reader.getNumericDocValues(VersionFieldMapper.NAME);
+        if (versions == null) {
+            throw new IllegalArgumentException("reader misses the [" + VersionFieldMapper.NAME +
+                "] field");
+        }
+        boolean assertionsOn = false;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java]: maybe:
```Java
Object readerKey = null;
assert (readerKey = reader.getCoreCacheKey()) != null;
this.readerKey = readerKey;
```

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 13]
```diff
@@ -675,13 +729,70 @@ private static void index(final List<ParseContext.Document> docs, final IndexWri
         }
     }
 
+    private static final class IndexingPlan {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I am not sure about the name, maybe `IndexingStrategy`?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 14]
```diff
@@ -610,24 +583,95 @@ public IndexResult index(Index index) throws IOException {
...
+            assert index.versionType().versionTypeForReplicationAndRecovery() == index.versionType() :
+                "resolving out of order delivery based on versioning but version type isn't fit for it. got ["
+                    + index.versionType() + "]";
+            // unlike the primary, replicas don't really care to about creation status of documents
+            // this allows to ignore the case where a document was found in the live version maps in
+            // a delete state and return false for the created flag in favor of code simplicity
+            final LuceneDocStatus luceneOpStatus = checkLuceneDocStatusBasedOnVersions(index);
+            if (luceneOpStatus == LuceneDocStatus.NEWER_OR_EQUAL) {
+                plan = IndexingPlan.processButSkipLucene(
+                    luceneOpStatus == LuceneDocStatus.NOT_FOUND, index.seqNo(), index.version());
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: `luceneOpStatus == LuceneDocStatus.NOT_FOUND` is `false` here?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 15]
```diff
@@ -610,24 +583,95 @@ public IndexResult index(Index index) throws IOException {
...
+            plan = IndexingPlan.optimizedAppendOnly(index.seqNo());
+        } else {
+            // drop out of order operations
+            assert index.versionType().versionTypeForReplicationAndRecovery() == index.versionType() :
+                "resolving out of order delivery based on versioning but version type isn't fit for it. got ["
+                    + index.versionType() + "]";
+            // unlike the primary, replicas don't really care to about creation status of documents
+            // this allows to ignore the case where a document was found in the live version maps in
+            // a delete state and return false for the created flag in favor of code simplicity
+            final LuceneDocStatus luceneOpStatus = checkLuceneDocStatusBasedOnVersions(index);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: so if this returns `LuceneDocStatus.NEWER_OR_EQUAL` then the existing doc is newer or equal and not the given doc? that is very confusing, I mentioned this before I think. I'd never expect that. The opposite is intuitive?

**Review by s1monw — CHANGES_REQUESTED** [review]
mostly nit picks looks great

File: core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java (hunk 12 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDAndVersionLookup.java]: much better. Thanks.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 14 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: yes! IntelliJ agrees too.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 13 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Sure.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 15 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Ok. I flipped it around.

**Comment by bleskes** [comment]
@s1monw I addressed your latest feedback (thx). Can you take another look please?

**Review by s1monw — APPROVED** [review]
LGTM thanks for the iterations

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by bleskes** [comment]
PS. I will let this bake for a day or two before back porting.
